server/core/src/https/mod.rs

```diff
@@ -429,568 +128,306 @@ pub async fn create_https_server(
+        )
+        // worker-src 'none';
+        .add(CspDirectiveType::WorkerSource, vec![CspValue::None])
+        // frame-ancestors 'none'
+        .add(CspDirectiveType::FrameAncestors, vec![CspValue::None])
+        .add(CspDirectiveType::ScriptSource, js_directives)
+        .add(
+            CspDirectiveType::ImgSrc,
+            vec![CspValue::SelfSite, CspValue::SchemeData],
+        );
+
+    // TODO this whole session store is kinda cursed and doesn't work the way we need, I think?
+    let store = async_session::CookieStore::new();
+    // let secret = b"..."; // MUST be at least 64 bytes!
+    let secret = format!("{:?}", cookie_key);
+    let secret = secret.as_bytes(); // TODO the cookie/session secret needs to be longer?
+    let session_layer = SessionLayer::new(store, secret)
+        .with_cookie_name("kanidm-session")
+        .with_session_ttl(None)
+        .with_cookie_domain(config.domain)
+        .with_same_site_policy(SameSite::Strict)
+        .with_secure(true);
+
+    let state = ServerState {
         status_ref,
         qe_w_ref,
         qe_r_ref,
         jws_signer,
         jws_validator,
-        js_files: js_files.to_owned(),
-        trust_x_forward_for,
-    });
-
-    // Add the logging subsystem.
-    tserver.with(sketching::middleware::TreeMiddleware::new(
-        trust_x_forward_for,
-    ));
-
-    // Add cookie handling.
-    tserver.with(
-        // We do not force a session ttl, because we validate this elsewhere in usage.
-        tide::sessions::SessionMiddleware::new(tide::sessions::CookieStore::new(), cookie_key)
-            .with_session_ttl(None)
-            .with_cookie_name("kanidm-session")
-            // Without this, the cookies won't be used on subdomains of origin.
-            .with_cookie_domain(domain)
-            // Im not sure if we need Lax here, I don't think we do because on the first get
-            // we don't need the cookie since wasm drives the fetches.
-            .with_same_site_policy(tide::http::cookies::SameSite::Strict),
-    );
-
-    // Strict responses.
-    tserver.with(StrictResponseMiddleware::default());
-
-    // Add routes
-    // ==== static content routes that have a longer cache policy.
+        js_files,
+        csp_header: csp_header.finish(),
+    };
 
-    // If we are no-ui, we remove this.
-    if !matches!(role, ServerRole::WriteReplicaNoUI) {
-        let pkg_path = PathBuf::from(env!("KANIDM_WEB_UI_PKG_PATH"));
-        if !pkg_path.exists() {
-            eprintln!(
-                "Couldn't find Web UI package path: ({}), quitting.",
-                env!("KANIDM_WEB_UI_PKG_PATH")
-            );
-            std::process::exit(1);
+    let static_routes = match config.role {
+        ServerRole::WriteReplica | ServerRole::ReadOnlyReplica => {
+            Router::new()
+                .route("/", get(crate::https::ui::ui_handler))
+                .route("/*ui", get(crate::https::ui::ui_handler))
```

Comment:
This is matching "/*ui" which also is /helloui and /tui. So I don't think this is the right url here. Are you trying to do the spa url capture? Because if you are you need:

```
    // This isn't clear but it means "serve this dir" followed by
    // "if nothing is found in the dir, serve pkg/index.html", it's not relative to /pkg.
    // So this will serve /hello too.
    let serve_dir = ServeDir::new("pkg")
        .not_found_service(ServeFile::new("pkg/index.html"));

    let spa_serve = Router::new()
        .nest_service("/pkg", serve_dir.clone())
        .fallback_service(serve_dir)
        .with_state(app_state.clone());
```

Reply:
This does the equivalent of a regex of `\/(?P<ui>.*)` and lets you add a "ui" `Path` extractor in `ui_handler` to catch the part of the path - ref [Wildcards](https://docs.rs/axum/0.6.18/axum/routing/struct.Router.html#wildcards).

Adding a `ServeFile` under `/ui/` isn't needed since all the static things are under `/pkg` and that's already there later based on ServerRole. `crate::https::ui::ui_handler` responds with the HTML which loads the WASM frontend.

I'm updating the routes to match /ui/* better - to `/ui/*key`, and a redirector from `/` to `/ui/login`.

Reply:
... and that explains why some of the tests were passing - it was responding with a 200 to /*